File: probes/receipt-envelope-encoder-v0/ts/canonical_cbor.ts
// Canonical CBOR encoder/decoder (hand-rolled, no dep).
//
// Implements RFC 8949 §4.2.1 deterministic encoding restricted to the subset
// declared in ../SPEC.md. Floats, indefinite-length, tags, undefined, bignums
// beyond u64/i64, non-string map keys, and non-canonical encodings are all
// forbidden — encoder throws, decoder rejects.

export type CborValue =
  | number
  | bigint
  | string
  | Uint8Array
  | boolean
  | null
  | CborValue[]
  | { [key: string]: CborValue };

// ────────────────────────────────────────────────────────────────────────
// ENCODER
// ────────────────────────────────────────────────────────────────────────

export function encodeCanonical(value: CborValue): Uint8Array {
  const out: number[] = [];
  encodeValue(value, out);
  return new Uint8Array(out);
}

function encodeValue(v: CborValue, out: number[]): void {
  if (v === null) { out.push(0xf6); return; }
  if (v === true) { out.push(0xf5); return; }
  if (v === false) { out.push(0xf4); return; }

  if (typeof v === "number") {
    if (!Number.isFinite(v)) throw new Error("canonical-cbor: non-finite number forbidden");
    if (!Number.isInteger(v)) throw new Error("canonical-cbor: floats forbidden (use Q-format integers)");
    if (!Number.isSafeInteger(v)) throw new Error("canonical-cbor: integer out of safe range; use bigint");
    encodeIntNum(v, out);
    return;
  }

  if (typeof v === "bigint") {
    encodeIntBig(v, out);
    return;
  }

  if (typeof v === "string") {
    const bytes = new TextEncoder().encode(v);
    encodeHead(3, bytes.length, out);
    pushAll(out, bytes);
    return;
  }

  if (v instanceof Uint8Array) {
    encodeHead(2, v.length, out);
    pushAll(out, v);
    return;
  }

  if (Array.isArray(v)) {
    encodeHead(4, v.length, out);
    for (const item of v) encodeValue(item, out);
    return;
  }

  if (typeof v === "object" && v !== null) {
    if (v instanceof Map) throw new Error("canonical-cbor: JS Map forbidden; use plain object");
    if (v instanceof Set) throw new Error("canonical-cbor: JS Set forbidden");
    const obj = v as Record<string, CborValue>;
    const keys = Object.keys(obj);

    // RFC 8949 §4.2.1: keys sorted by bytewise lexicographic order of their
    // deterministic encodings.
    const encoded: { key: string; bytes: Uint8Array }[] = keys.map((k) => {
      const tmp: number[] = [];
      const kb = new TextEncoder().encode(k);
      encodeHead(3, kb.length, tmp);
      pushAll(tmp, kb);
      return { key: k, bytes: new Uint8Array(tmp) };
    });
    encoded.sort((a, b) => compareBytes(a.bytes, b.bytes));

    for (let i = 1; i < encoded.length; i++) {
      if (compareBytes(encoded[i - 1].bytes, encoded[i].bytes) === 0) {
        throw new Error(`canonical-cbor: duplicate map key after encoding: ${encoded[i].key}`);
      }
    }

    encodeHead(5, encoded.length, out);
    for (const { key, bytes } of encoded) {
      pushAll(out, bytes);
      encodeValue(obj[key], out);
    }
    return;
  }

  throw new Error(`canonical-cbor: unsupported value type: ${typeof v}`);
}

function encodeIntNum(n: number, out: number[]): void {
  // Safe-integer range goes up to 2^53-1, which exceeds u32. For values
  // beyond u32, delegate to bigint encoder (which uses 8-byte form).
  if (n >= 0 && n < 0x100000000) { encodeHead(0, n, out); return; }
  if (n < 0 && -1 - n < 0x100000000) { encodeHead(1, -1 - n, out); return; }
  encodeIntBig(BigInt(n), out);
}

function encodeIntBig(n: bigint, out: number[]): void {
  // 8-byte form covers everything from 2^32 up to u64. encodeHead handles
  // anything below 2^32. Both forms are shortest-canonical for their range.
  if (n >= 0n) {
    if (n < 0x100000000n) {
      encodeHead(0, Number(n), out);
      return;
    }
    if (n > 0xffffffffffffffffn) throw new Error("canonical-cbor: uint > u64 forbidden");
    out.push(0x1b);
    for (let i = 7; i >= 0; i--) out.push(Number((n >> BigInt(i * 8)) & 0xffn));
    return;
  }
  const negated = -1n - n;
  if (negated < 0x100000000n) {
    encodeHead(1, Number(negated), out);
    return;
  }
  if (negated > 0xffffffffffffffffn) throw new Error("canonical-cbor: negint < -2^64 forbidden");
  out.push(0x3b);
  for (let i = 7; i >= 0; i--) out.push(Number((negated >> BigInt(i * 8)) & 0xffn));
}

function encodeHead(major: number, value: number, out: number[]): void {
  const m = (major << 5) & 0xff;
  if (value < 0) throw new Error("canonical-cbor: negative head argument");
  if (value < 24) {
    out.push(m | value);
  } else if (value < 0x100) {
    out.push(m | 24, value & 0xff);
  } else if (value < 0x10000) {
    out.push(m | 25, (value >> 8) & 0xff, value & 0xff);
  } else if (value < 0x100000000) {
    out.push(m | 26, (value >>> 24) & 0xff, (value >> 16) & 0xff, (value >> 8) & 0xff, value & 0xff);
  } else {
    throw new Error("canonical-cbor: value too large; use bigint for >= 2^32");
  }
}

function pushAll(out: number[], bytes: Uint8Array): void {
  for (let i = 0; i < bytes.length; i++) out.push(bytes[i]);
}

function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

// ────────────────────────────────────────────────────────────────────────
// DECODER (strict canonical: any successful decode re-encodes to same bytes)
// ────────────────────────────────────────────────────────────────────────

type DecodeCtx = { buf: Uint8Array; pos: number };

export function decodeCanonical(bytes: Uint8Array): CborValue {
  const ctx: DecodeCtx = { buf: bytes, pos: 0 };
  const v = decodeValue(ctx);
  if (ctx.pos !== bytes.length) {
    throw new Error(`canonical-cbor: trailing bytes (${bytes.length - ctx.pos} unread)`);
  }
  return v;
}

function decodeValue(ctx: DecodeCtx): CborValue {
  const b = readByte(ctx);
  const major = (b >> 5) & 0x07;
  const info = b & 0x1f;

  if (info === 31) throw new Error("canonical-cbor: indefinite-length form forbidden");

  // major 7: simple values + floats
  if (major === 7) {
    if (info === 20) return false;
    if (info === 21) return true;
    if (info === 22) return null;
    if (info === 23) throw new Error("canonical-cbor: undefined (simple 23) forbidden");
    if (info === 25 || info === 26 || info === 27) {
      throw new Error("canonical-cbor: floating-point forbidden");
    }
    throw new Error(`canonical-cbor: simple value ${info} forbidden`);
  }

  // major 6: tags
  if (major === 6) throw new Error("canonical-cbor: tags forbidden");

  // length argument for majors 0..5 — must be shortest form
  const value = readArgCanonical(ctx, info);

  switch (major) {
    case 0:
      return value <= Number.MAX_SAFE_INTEGER ? Number(value) : value;
    case 1: {
      // -1 - value
      const v = value;
      if (typeof v === "bigint") return -1n - v;
      return -1 - v;
    }
    case 2: {
      const len = toLen(value);
      const bytes = ctx.buf.slice(ctx.pos, ctx.pos + len);
      if (bytes.length !== len) throw new Error("canonical-cbor: truncated byte string");
      ctx.pos += len;
      return bytes;
    }
    case 3: {
      const len = toLen(value);
      const bytes = ctx.buf.slice(ctx.pos, ctx.pos + len);
      if (bytes.length !== len) throw new Error("canonical-cbor: truncated text string");
      ctx.pos += len;
      return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    }
    case 4: {
      const len = toLen(value);
      const arr: CborValue[] = [];
      for (let i = 0; i < len; i++) arr.push(decodeValue(ctx));
      return arr;
    }
    case 5: {
      const len = toLen(value);
      const obj: Record<string, CborValue> = {};
      let prevKeyBytes: Uint8Array | null = null;
      for (let i = 0; i < len; i++) {
        const keyStartPos = ctx.pos;
        const key = decodeValue(ctx);
        if (typeof key !== "string") {
          throw new Error("canonical-cbor: non-string map key forbidden");
        }
        const keyBytes = ctx.buf.slice(keyStartPos, ctx.pos);
        if (prevKeyBytes !== null && compareBytes(prevKeyBytes, keyBytes) >= 0) {
          throw new Error("canonical-cbor: map keys not in bytewise-lex order");
        }
        prevKeyBytes = keyBytes;
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          throw new Error(`canonical-cbor: duplicate map key: ${key}`);
        }
        obj[key] = decodeValue(ctx);
      }
      return obj;
    }
  }

  throw new Error(`canonical-cbor: unreachable (major ${major})`);
}

function readByte(ctx: DecodeCtx): number {
  if (ctx.pos >= ctx.buf.length) throw new Error("canonical-cbor: unexpected end");
  return ctx.buf[ctx.pos++];
}

function readArgCanonical(ctx: DecodeCtx, info: number): number | bigint {
  if (info < 24) return info;
  if (info === 24) {
    const v = readByte(ctx);
    if (v < 24) throw new Error("canonical-cbor: non-canonical (1-byte length < 24)");
    return v;
  }
  if (info === 25) {
    const v = (readByte(ctx) << 8) | readByte(ctx);
    if (v < 0x100) throw new Error("canonical-cbor: non-canonical (2-byte length < 256)");
    return v;
  }
  if (info === 26) {
    const v = (readByte(ctx) * 0x1000000) + (readByte(ctx) << 16) + (readByte(ctx) << 8) + readByte(ctx);
    if (v < 0x10000) throw new Error("canonical-cbor: non-canonical (4-byte length < 65536)");
    return v;
  }
  if (info === 27) {
    let v = 0n;
    for (let i = 0; i < 8; i++) v = (v << 8n) | BigInt(readByte(ctx));
    if (v < 0x100000000n) throw new Error("canonical-cbor: non-canonical (8-byte length < 2^32)");
    return v;
  }
  throw new Error(`canonical-cbor: invalid additional info ${info}`);
}

function toLen(v: number | bigint): number {
  if (typeof v === "bigint") {
    if (v > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new Error("canonical-cbor: length exceeds safe integer range");
    }
    return Number(v);
  }
  return v;
}

// ────────────────────────────────────────────────────────────────────────
// HASHING (multihash sha-256 as hex string)
// ────────────────────────────────────────────────────────────────────────

export async function multihashSha256(bytes: Uint8Array): Promise<string> {
  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", bytes));
  const hex = Array.from(digest, (b) => b.toString(16).padStart(2, "0")).join("");
  return "1220" + hex;
}

export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

File: probes/receipt-envelope-encoder-v0/python/canonical_cbor.py
"""Canonical CBOR encoder/decoder (hand-rolled, stdlib-only).

Faithful port of ../ts/canonical_cbor.ts. Implements RFC 8949 §4.2.1
deterministic encoding restricted to the subset declared in ../SPEC.md.

Forbidden constructs match the TS impl exactly:
  - floats (encoder rejects; decoder rejects)
  - indefinite-length forms
  - tags
  - undefined / None as value (None is allowed only as null)
  - bignums > u64 (positive) or < -2^64 (negative)
  - non-string map keys
  - duplicate keys after encoding
  - trailing bytes after decode
  - non-canonical re-encodings (rejected at decode time)

Second-implementation gate for RECEIPT_ENVELOPE.v0.1 (per Codex review
2026-05-14T173027Z): bytes from this impl MUST match bytes from the TS
impl for the same input.
"""

from __future__ import annotations

import hashlib
from typing import Any, List, Tuple, Union

CborValue = Union[
    int,
    str,
    bytes,
    bytearray,
    bool,
    None,
    List["CborValue"],
    dict,
]


# ────────────────────────────────────────────────────────────────────────
# ENCODER
# ────────────────────────────────────────────────────────────────────────


def encode_canonical(value: Any) -> bytes:
    out = bytearray()
    _encode_value(value, out)
    return bytes(out)


def _encode_value(v: Any, out: bytearray) -> None:
    # Order of type checks matters: bool MUST come before int (in Python,
    # isinstance(True, int) is True).
    if v is None:
        out.append(0xF6)
        return
    if v is True:
        out.append(0xF5)
        return
    if v is False:
        out.append(0xF4)
        return

    if isinstance(v, int):
        if v >= 0:
            if v > 0xFFFFFFFFFFFFFFFF:
                raise ValueError("canonical-cbor: uint > u64 forbidden")
            _encode_int(0, v, out)
        else:
            negated = -1 - v
            if negated > 0xFFFFFFFFFFFFFFFF:
                raise ValueError("canonical-cbor: negint < -2^64 forbidden")
            _encode_int(1, negated, out)
        return

    if isinstance(v, float):
        raise ValueError(
            "canonical-cbor: floats forbidden (use Q-format integers)"
        )

    if isinstance(v, str):
        b = v.encode("utf-8")
        _encode_head(3, len(b), out)
        out.extend(b)
        return

    if isinstance(v, (bytes, bytearray)):
        _encode_head(2, len(v), out)
        out.extend(v)
        return

    if isinstance(v, list):
        _encode_head(4, len(v), out)
        for item in v:
            _encode_value(item, out)
        return

    if isinstance(v, tuple):
        # Tuples encode as arrays — same as lists.
        _encode_head(4, len(v), out)
        for item in v:
            _encode_value(item, out)
        return

    if isinstance(v, dict):
        # RFC 8949 §4.2.1: sort by bytewise-lex order of encoded keys.
        encoded_keys: List[Tuple[Any, bytes]] = []
        for k in v.keys():
            if not isinstance(k, str):
                raise ValueError(
                    "canonical-cbor: non-string map key forbidden"
                )
            kb = k.encode("utf-8")
            head = bytearray()
            _encode_head(3, len(kb), head)
            encoded_keys.append((k, bytes(head) + kb))

        encoded_keys.sort(key=lambda kv: kv[1])

        # Check for duplicates after encoding.
        for i in range(1, len(encoded_keys)):
            if encoded_keys[i][1] == encoded_keys[i - 1][1]:
                raise ValueError(
                    f"canonical-cbor: duplicate map key: {encoded_keys[i][0]}"
                )

        _encode_head(5, len(encoded_keys), out)
        for key, key_bytes in encoded_keys:
            out.extend(key_bytes)
            _encode_value(v[key], out)
        return

    raise ValueError(f"canonical-cbor: unsupported value type: {type(v).__name__}")


def _encode_int(major: int, value: int, out: bytearray) -> None:
    """Encode major-type 0 or 1 integer using shortest canonical form."""
    if value < 0:
        raise ValueError("canonical-cbor: negative head argument")
    if value < 0x100000000:
        _encode_head(major, value, out)
        return
    # 8-byte form for values 2^32 .. 2^64-1
    out.append(((major & 0x07) << 5) | 27)
    for i in range(7, -1, -1):
        out.append((value >> (i * 8)) & 0xFF)


def _encode_head(major: int, value: int, out: bytearray) -> None:
    """Encode major-type and value header using shortest form for value < 2^32."""
    if value < 0:
        raise ValueError("canonical-cbor: negative head argument")
    m = (major << 5) & 0xFF
    if value < 24:
        out.append(m | value)
    elif value < 0x100:
        out.append(m | 24)
        out.append(value & 0xFF)
    elif value < 0x10000:
        out.append(m | 25)
        out.append((value >> 8) & 0xFF)
        out.append(value & 0xFF)
    elif value < 0x100000000:
        out.append(m | 26)
        out.append((value >> 24) & 0xFF)
        out.append((value >> 16) & 0xFF)
        out.append((value >> 8) & 0xFF)
        out.append(value & 0xFF)
    else:
        raise ValueError(
            "canonical-cbor: value too large for 4-byte head; use 8-byte path"
        )


# ────────────────────────────────────────────────────────────────────────
# DECODER (strict canonical)
# ────────────────────────────────────────────────────────────────────────


class _DecodeCtx:
    def __init__(self, buf: bytes) -> None:
        self.buf = buf
        self.pos = 0


def decode_canonical(data: bytes) -> Any:
    ctx = _DecodeCtx(data)
    v = _decode_value(ctx)
    if ctx.pos != len(data):
        raise ValueError(
            f"canonical-cbor: trailing bytes ({len(data) - ctx.pos} unread)"
        )
    return v


def _decode_value(ctx: _DecodeCtx) -> Any:
    b = _read_byte(ctx)
    major = (b >> 5) & 0x07
    info = b & 0x1F

    if info == 31:
        raise ValueError("canonical-cbor: indefinite-length form forbidden")

    if major == 7:
        if info == 20:
            return False
        if info == 21:
            return True
        if info == 22:
            return None
        if info == 23:
            raise ValueError("canonical-cbor: undefined (simple 23) forbidden")
        if info in (25, 26, 27):
            raise ValueError("canonical-cbor: floating-point forbidden")
        raise ValueError(f"canonical-cbor: simple value {info} forbidden")

    if major == 6:
        raise ValueError("canonical-cbor: tags forbidden")

    value = _read_arg_canonical(ctx, info)

    if major == 0:
        return value
    if major == 1:
        return -1 - value
    if major == 2:
        n = value
        bs = ctx.buf[ctx.pos:ctx.pos + n]
        if len(bs) != n:
            raise ValueError("canonical-cbor: truncated byte string")
        ctx.pos += n
        return bytes(bs)
    if major == 3:
        n = value
        bs = ctx.buf[ctx.pos:ctx.pos + n]
        if len(bs) != n:
            raise ValueError("canonical-cbor: truncated text string")
        ctx.pos += n
        return bs.decode("utf-8")
    if major == 4:
        n = value
        return [_decode_value(ctx) for _ in range(n)]
    if major == 5:
        n = value
        obj = {}
        prev_key_bytes = None
        for _ in range(n):
            key_start = ctx.pos
            key = _decode_value(ctx)
            if not isinstance(key, str):
                raise ValueError("canonical-cbor: non-string map key forbidden")
            key_bytes = ctx.buf[key_start:ctx.pos]
            if prev_key_bytes is not None and key_bytes <= prev_key_bytes:
                raise ValueError(
                    "canonical-cbor: map keys not in bytewise-lex order"
                )
            prev_key_bytes = key_bytes
            if key in obj:
                raise ValueError(f"canonical-cbor: duplicate map key: {key}")
            obj[key] = _decode_value(ctx)
        return obj

    raise ValueError(f"canonical-cbor: unreachable (major {major})")


def _read_byte(ctx: _DecodeCtx) -> int:
    if ctx.pos >= len(ctx.buf):
        raise ValueError("canonical-cbor: unexpected end")
    b = ctx.buf[ctx.pos]
    ctx.pos += 1
    return b


def _read_arg_canonical(ctx: _DecodeCtx, info: int) -> int:
    if info < 24:
        return info
    if info == 24:
        v = _read_byte(ctx)
        if v < 24:
            raise ValueError("canonical-cbor: non-canonical (1-byte length < 24)")
        return v
    if info == 25:
        hi = _read_byte(ctx)
        lo = _read_byte(ctx)
        v = (hi << 8) | lo
        if v < 0x100:
            raise ValueError(
                "canonical-cbor: non-canonical (2-byte length < 256)"
            )
        return v
    if info == 26:
        v = 0
        for _ in range(4):
            v = (v << 8) | _read_byte(ctx)
        if v < 0x10000:
            raise ValueError(
                "canonical-cbor: non-canonical (4-byte length < 65536)"
            )
        return v
    if info == 27:
        v = 0
        for _ in range(8):
            v = (v << 8) | _read_byte(ctx)
        if v < 0x100000000:
            raise ValueError(
                "canonical-cbor: non-canonical (8-byte length < 2^32)"
            )
        return v
    raise ValueError(f"canonical-cbor: invalid additional info {info}")


# ────────────────────────────────────────────────────────────────────────
# HASHING
# ────────────────────────────────────────────────────────────────────────


def multihash_sha256(data: bytes) -> str:
    digest = hashlib.sha256(data).digest()
    return "1220" + digest.hex()


def to_hex(data: bytes) -> str:
    return data.hex()
